developer/src/kmldmlc/src/keyman/compiler/vkey.ts

```diff
@@ -0,0 +1,68 @@
+import { constants } from "@keymanapp/ldml-keyboard-constants";
+import { Vkey } from "../kmx/kmx-plus";
+import { LdmlVkeyNames } from "../ldml-keyboard/us-virtual-keys";
+import { CompilerMessages } from "./messages";
+import { SectionCompiler } from "./section-compiler";
+
+export class VkeyCompiler extends SectionCompiler {
+
+  public get id() {
+    return constants.section.vkey;
+  }
+
+  public get required(): boolean {
+    return !!this.keyboard.vkeyMaps;
+  }
+
+  public validate(): boolean {
+    let valid = true;
+    if(this.keyboard.vkeyMaps) {
+      let from: string[] = [], to: string[] = [];
+
+      this.keyboard.vkeyMaps.vkeyMap.forEach(vk => {
+        if(LdmlVkeyNames[vk.from] === undefined) {
+          this.callbacks.reportMessage(CompilerMessages.Error_VkeyIsNotValid(vk.from));
+          valid = false;
+        }
+
+        if(LdmlVkeyNames[vk.to] === undefined) {
+          this.callbacks.reportMessage(CompilerMessages.Error_VkeyIsNotValid(vk.to));
+          valid = false;
+        }
+
+        if(vk.from == vk.to) {
+          this.callbacks.reportMessage(CompilerMessages.Hint_VkeyMapIsRedundant(vk.from));
+        }
+
+        if(from.find(svk => svk == vk.from)) {
+          this.callbacks.reportMessage(CompilerMessages.Error_VkeyMapIsRepeated(vk.from));
+          valid = false;
+        }
+        from.push(vk.from);
+
+        if(to.find(svk => svk == vk.to)) {
+          this.callbacks.reportMessage(CompilerMessages.Info_MultipleVkeyMapsHaveSameTarget(vk.to));
```

Comment:
Is this really an error? Are the vkey maps supposed to be 1:1?

Reply:
I noted it as an info message because I thought there's a good chance that two VKs producing the same output for hardware keyboards is going to be wrong (of course, 102nd key is same as backslash on US on Windows so YMMV). Idea is that info messages can be suppressed individually or en masse in the future.